linuxrc.c

```diff
@@ -1567,3 +1588,71 @@ void lxrc_readd_parts()
 }
 
 
+/*
+ * Offer the user a list of URLs to choose from based in the current install
+ * URL and the default repo setting.
+ *
+ * In addidtion she can also edit the URL.
+ *
+ * Return NULL in repo if the user cancelled the dialogs, else repo contains
+ * the chosen repo URL.
+ */
+void select_repo_url(char *msg, char **repo)
+{
+  slist_t *sl;
+  char **item_list;
+  int i, with_install_url, items, default_item;
+
+  // if config.url.install is a duplicate of some default repo entry, skip it
+  with_install_url = config.url.install && !slist_getentry(config.defaultrepo, config.url.install->str);
+
+  for(items = 1, sl = config.defaultrepo; sl; sl = sl->next) items++;
+
+  if(with_install_url) items++;
+
+  item_list = calloc(items + 1, sizeof *item_list);
+
+  i = 0;
+
+  if(with_install_url) {
+    item_list[i++] = config.url.install->str;
+  }
+
+  for(sl = config.defaultrepo; sl; sl = sl->next, i++) {
+    item_list[i] = sl->key;
+  }
+
+  item_list[i] = "Enter another URL";
+
+  default_item = 1;
+
+  // if we skipped config.url.install because it's a duplicate of some
+  // defaultrepo entry, make it the default entry
+  if(config.url.install && !with_install_url) {
+    for(i = 1, sl = config.defaultrepo; sl && strcmp(sl->key, config.url.install->str); sl = sl->next) i++;
+    if(sl) default_item = i;
+  }
+
+  i = dia_list(msg, 64, NULL, item_list, default_item, align_left);
+
+  if(i > 0) {
+    if(i == items) {
+      if(!*repo) str_copy(repo, item_list[default_item - 1]);
+      if(dia_input2("Enter the repository URL.", repo, 64, 0)) {
```

Comment:
is 64 hard limit or soft width of text entry? As I worry that there can be longer URL.